File: 396.rotate-function.py
#
# @lc app=leetcode id=396 lang=python3
#
# [396] Rotate Function
#

# Time:  O(n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def maxRotateFunction(self, A: List[int]) -> int:
        s = sum(A)
        fi = 0
        for i in range(len(A)):
            fi += i * A[i]

        # F(i) = F(i-1) + sum - n * A[n-1]
        res = fi
        for i in range(1, len(A) + 1):
            fi += s - len(A) * A[-i]
            res = max(res, fi)
        return res


# @lc code=end

File: 396.rotate-function.js
/*
 * @lc app=leetcode id=396 lang=javascript
 *
 * [396] Rotate Function
 */

// @lc code=start
/**
 * @param {number[]} A
 * @return {number}
 */
var maxRotateFunction = function (A) {
  const sum = A.reduce((accu, curr) => accu + curr);
  let fi = 0;
  for (let i = 0; i < A.length; ++i) {
    fi += i * A[i];
  }
  let res = fi;
  for (let i = 0; i < A.length; ++i) {
    fi += sum - A.length * A[A.length - 1 - i];
    res = Math.max(res, fi);
  }
  return res;
};
// @lc code=end
